File: decompiled/c/5e3bbf.c

short FUN_005e3bbf(void)

{
  short in_AX;
  short in_CX;
  short in_DX;
  short unaff_BX;
  undefined *puVar1;
  
  if ((((-1 < in_DX) && (0x1d < in_AX)) && ((short)(in_DX + unaff_BX) <= DAT_00971ed6)) &&
     ((short)(in_AX + in_CX) <= DAT_00971ed8)) {
    puVar1 = &DAT_009a013c;
    while( true ) {
      if (DAT_009a1164 <= puVar1) {
        return in_AX;
      }
      if ((((*(ushort *)(puVar1 + 0x32) & 1) == 0) &&
          (*(short *)(puVar1 + 0x20) < (short)(in_DX + unaff_BX))) &&
         ((in_DX < (short)(*(short *)(puVar1 + 0x20) + *(short *)(puVar1 + 0x24)) &&
          ((*(short *)(puVar1 + 0x22) < (short)(in_AX + in_CX) &&
           (in_AX < (short)(*(short *)(puVar1 + 0x22) + *(short *)(puVar1 + 0x26)))))))) break;
      puVar1 = puVar1 + 0x178;
    }
  }
  return in_AX;
}

File: ported/auto/5e3bbf.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/5e3bbf.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { regs } from "../../runtime/regs.js";
export function FUN_005e3bbf(heap) {
  let in_AX = regs.eax & 0xffff;
  let in_CX = regs.ecx & 0xffff;
  let in_DX = regs.edx & 0xffff;
  let unaff_BX = regs.ebx & 0xffff;
  let puVar1 = 0;
  if ((((-1 < (in_DX | 0)) && (0x1d < in_AX)) && ((((in_DX + unaff_BX)) << 16 >> 16) <= heap.u32(0x00971ed6))) && ((((in_AX + in_CX)) << 16 >> 16) <= heap.u32(0x00971ed8))) {
    puVar1 = ((0x009a013c) >>> 0);
    while (true) {
      if (heap.u32(0x009a1164) <= puVar1) {
        return in_AX;
      }
      if ((((heap.u16((puVar1 + 0x32)) & 1) == 0) && (heap.i16((puVar1 + 0x20)) < (((in_DX + unaff_BX)) << 16 >> 16))) && ((in_DX < (((heap.i16((puVar1 + 0x20)) + heap.i16((puVar1 + 0x24)))) << 16 >> 16) && ((heap.i16((puVar1 + 0x22)) < (((in_AX + in_CX)) << 16 >> 16) && (in_AX < (((heap.i16((puVar1 + 0x22)) + heap.i16((puVar1 + 0x26)))) << 16 >> 16))))))) {
        break;
      }
      puVar1 = ((puVar1 + 0x178) >>> 0);
    }
  }
  return in_AX;
}
